diff --git a/tools/coverage/report_generator.sh.template b/tools/coverage/report_generator.sh.template
index 631ec26361..faab364644 100755
--- a/tools/coverage/report_generator.sh.template
+++ b/tools/coverage/report_generator.sh.template
@@ -107,7 +107,7 @@ run_grcov() {
         "${INFO_FILES[@]}" \
         --precision 1 \
         -s "${WORKSPACE_PATH}" \
-        -t html,covdir \
+        -t html,covdir,lcov \
         --excl-line "LCOV_EXCL_LINE" \
         --output-config-file grcov_config_updated.json \
         -o "${OUTPUT_DIR}"
