diff --git a/test/integration/server.cc b/test/integration/server.cc
index 903cb15c36..d5f9c8000a 100644
--- a/test/integration/server.cc
+++ b/test/integration/server.cc
@@ -298,7 +298,7 @@ IntegrationTestServerImpl::~IntegrationTestServerImpl() {
       server_gone_.WaitForNotification();
     }
   } else {
-    if (server_) {
+    if (!server_gone_.HasBeenNotified()) {
       server_->dispatcher().post([this]() { server_->shutdown(); });
       server_gone_.WaitForNotification();
     }
