diff --git a/source/extensions/filters/common/lua/BUILD b/source/extensions/filters/common/lua/BUILD
index 9b4cae8785..f69a536317 100644
--- a/source/extensions/filters/common/lua/BUILD
+++ b/source/extensions/filters/common/lua/BUILD
@@ -13,13 +13,13 @@ envoy_cc_library(
     srcs = ["lua.cc"],
     hdrs = ["lua.h"],
     deps = [
-        "//bazel/foreign_cc:luajit",
         "//envoy/thread_local:thread_local_interface",
         "//source/common/common:assert_lib",
         "//source/common/common:c_smart_ptr_lib",
         "//source/common/common:lock_guard_lib",
         "//source/common/common:thread_lib",
         "//source/common/protobuf",
+        "@luajit",
     ],
 )
 
@@ -41,10 +41,10 @@ envoy_cc_library(
     hdrs = ["protobuf_converter.h"],
     deps = [
         ":lua_lib",
-        "//bazel/foreign_cc:luajit",
         "//source/common/protobuf",
         "//source/common/protobuf:create_reflectable_message_lib",
         "//source/common/protobuf:utility_lib",
         "@abseil-cpp//absl/strings",
+        "@luajit",
     ],
 )
