diff --git a/bazel/envoy_test.bzl b/bazel/envoy_test.bzl
index 96bddca75a..fa09c8d112 100644
--- a/bazel/envoy_test.bzl
+++ b/bazel/envoy_test.bzl
@@ -186,7 +186,9 @@ def envoy_cc_test(
         srcs = srcs,
         data = data + select({
             "%s//bazel:local_asan_build" % repository: [],
-            "%s//bazel:asan_build" % repository: ["@llvm_toolchain_llvm//:symbolizer"],
+            "%s//bazel:asan_build" % repository: ["@llvm_toolchain//:llvm-symbolizer"],
+            "%s//bazel:msan_build" % repository: ["@llvm_toolchain//:llvm-symbolizer"],
+            "%s//bazel:tsan_build" % repository: ["@llvm_toolchain//:llvm-symbolizer"],
             "//conditions:default": [],
         }),
         copts = envoy_copts(repository, test = True) + copts + envoy_pch_copts(repository, "//test:test_pch"),
@@ -209,7 +211,12 @@ def envoy_cc_test(
         flaky = flaky,
         env = env | select({
             "%s//bazel:local_asan_build" % repository: {"ASAN_SYMBOLIZER_PATH": "%s/bin/llvm-symbolizer" % LLVM_PATH},
-            "%s//bazel:asan_build" % repository: {"ASAN_SYMBOLIZER_PATH": "$(location @llvm_toolchain_llvm//:symbolizer)"},
+            "%s//bazel:asan_build" % repository: {
+                "ASAN_SYMBOLIZER_PATH": "$(location @llvm_toolchain//:llvm-symbolizer)",
+                "UBSAN_SYMBOLIZER_PATH": "$(location @llvm_toolchain//:llvm-symbolizer)",
+            },
+            "%s//bazel:msan_build" % repository: {"MSAN_SYMBOLIZER_PATH": "$(location @llvm_toolchain//:llvm-symbolizer)"},
+            "%s//bazel:tsan_build" % repository: {"TSAN_SYMBOLIZER_PATH": "$(location @llvm_toolchain//:llvm-symbolizer)"},
             "//conditions:default": {},
         }),
         exec_properties = exec_properties,
