diff --git a/bazel/toolchains.bzl b/bazel/toolchains.bzl
index b02ca3bc57..c1c69e56fc 100644
--- a/bazel/toolchains.bzl
+++ b/bazel/toolchains.bzl
@@ -1,9 +1,8 @@
 load("@envoy_repo//:compiler.bzl", "LLVM_LIB_DIR", "LLVM_PATH", "LLVM_VERSION_LOCAL", "USE_LOCAL_SYSROOT")
 load("@envoy_toolshed//repository:utils.bzl", "arch_alias")
+load("@pomerium_envoy//bazel:toolchains.bzl", "LLVM_VERSION")
 load("@toolchains_llvm//toolchain:rules.bzl", "llvm_toolchain")

-_LLVM_VERSION_HERMETIC = "18.1.8"
-LLVM_VERSION = LLVM_VERSION_LOCAL if LLVM_VERSION_LOCAL else _LLVM_VERSION_HERMETIC
 LLVM_MAJOR = LLVM_VERSION.split(".")[0]
 LLVM_MAJOR_MINOR = ".".join(LLVM_VERSION.split(".")[:2])

