diff --git a/source/extensions/tracers/opentelemetry/BUILD b/source/extensions/tracers/opentelemetry/BUILD
index 0b6e75abea..0cac8092aa 100644
--- a/source/extensions/tracers/opentelemetry/BUILD
+++ b/source/extensions/tracers/opentelemetry/BUILD
@@ -24,4 +24,5 @@ envoy_cc_extension(

 envoy_cc_library(
+    visibility = ["//visibility:public"],
     name = "opentelemetry_tracer_lib",
     srcs = [