--- a/envoy.bazelrc
+++ b/envoy.bazelrc
@@ -156,15 +156,15 @@ build:asan-common --test_env=ASAN_OPTIONS=handle_abort=1:allow_addr2line=true:ch
 build:asan-common --test_env=UBSAN_OPTIONS=halt_on_error=true:print_stacktrace=1
 build:asan-common --test_env=ASAN_SYMBOLIZER_PATH
 # ASAN needs -O1 to get reasonable performance.
-build:asan-common --copt -O1
+build:asan-common --copt -O0
 build:asan-common --copt -fno-optimize-sibling-calls
 
 # ASAN config with clang runtime
 build:asan --config=asan-common
 build:asan --linkopt --rtlib=compiler-rt
 build:asan --linkopt --unwindlib=libgcc
-build:asan --linkopt=-l:libclang_rt.ubsan_standalone.a
-build:asan --linkopt=-l:libclang_rt.ubsan_standalone_cxx.a
+build:asan --linkopt=-l:libclang_rt.ubsan_standalone-x86_64.a
+build:asan --linkopt=-l:libclang_rt.ubsan_standalone_cxx-x86_64.a
 build:asan --action_env=ENVOY_UBSAN_VPTR=1
 build:asan --copt=-fsanitize=vptr,function
 build:asan --linkopt=-fsanitize=vptr,function
