diff --git a/.bazelrc b/.bazelrc
index c0bda60546..0221b5caf9 100644
--- a/.bazelrc
+++ b/.bazelrc
@@ -249,13 +249,11 @@ test:sanitizer --build_tests_only
 # ASAN config with clang runtime
 build:asan --config=asan-common
 build:asan --linkopt --rtlib=compiler-rt
-build:asan --linkopt --unwindlib=libgcc
 build:asan --linkopt=-l:libclang_rt.ubsan_standalone.a
 build:asan --linkopt=-l:libclang_rt.ubsan_standalone_cxx.a
 build:asan --action_env=ENVOY_UBSAN_VPTR=1
 build:asan --copt=-fsanitize=vptr,function
 build:asan --linkopt=-fsanitize=vptr,function
-build:asan --linkopt='-L/opt/llvm/lib/clang/18/lib/x86_64-unknown-linux-gnu'
 
 # Basic ASAN/UBSAN that works for gcc or llvm
 build:asan-common --config=sanitizer
@@ -276,7 +274,7 @@ build:asan-common --test_env=ASAN_OPTIONS=handle_abort=1:allow_addr2line=true:ch
 build:asan-common --test_env=UBSAN_OPTIONS=halt_on_error=true:print_stacktrace=1
 build:asan-common --test_env=ASAN_SYMBOLIZER_PATH
 # ASAN needs -O1 to get reasonable performance.
-build:asan-common --copt -O1
+build:asan-common --copt -O0
 build:asan-common --copt -fno-optimize-sibling-calls
 
 # macOS ASAN/UBSAN
