diff --git a/toolchain/cc_toolchain_config.bzl b/toolchain/cc_toolchain_config.bzl
index 06aae53..d059faf 100644
--- a/toolchain/cc_toolchain_config.bzl
+++ b/toolchain/cc_toolchain_config.bzl
@@ -234,6 +234,8 @@ def cc_toolchain_config(
     elif target_arch in ["wasm32", "wasm64"]:
         # lld is invoked as wasm-ld for WebAssembly targets.
         use_libtool = False
+    elif exec_os != "darwin":
+        use_libtool = False
     else:
         link_flags.extend([
             "-Wl,--build-id=md5",
@@ -294,6 +296,13 @@ def cc_toolchain_config(
                 "-lc++abi",
                 "-Bdynamic",
             ])
+        elif target_os == "darwin":
+            # Cross-compiling from linux to darwin
+            link_libs.extend([
+                "-lc++",
+                "-lc++abi",
+            ])
+            compiler_rt_link_flags = ["-rtlib=compiler-rt"]
         else:
             # For single-platform builds, we can statically link the bundled
             # libraries.
@@ -315,10 +324,12 @@ def cc_toolchain_config(
             "-stdlib=libc++",
         ]
 
-        link_libs.extend([
-            "-l:libc++.a",
-            "-l:libc++abi.a",
-        ])
+        compiler_rt_link_flags = ["-rtlib=compiler-rt"]
+        if target_os != "darwin":
+            link_libs.extend([
+                "-l:libc++.a",
+                "-l:libc++abi.a",
+            ])
     elif stdlib == "dynamic-stdc++":
         cxx_flags = [
             "-std=" + cxx_standard,
@@ -341,12 +352,14 @@ def cc_toolchain_config(
         cxx_flags = [
             "-std=" + cxx_standard,
         ]
-        libunwind_link_flags = [
-            "-l:libunwind.a",
-            # To support libunwind.
-            "-lpthread",
-            "-ldl",
-        ]
+        if target_os != "darwin":
+            compiler_rt_link_flags = ["-rtlib=compiler-rt"]
+            libunwind_link_flags = [
+                "-l:libunwind.a",
+                # To support libunwind.
+                "-lpthread",
+                "-ldl",
+            ]
     elif stdlib == "none":
         cxx_flags = [
             "-nostdlib",
diff --git a/toolchain/internal/configure.bzl b/toolchain/internal/configure.bzl
index 8503773..7e19db1 100644
--- a/toolchain/internal/configure.bzl
+++ b/toolchain/internal/configure.bzl
@@ -158,6 +158,7 @@ def llvm_config_impl(rctx):
     for target_pair, cxx_cross_lib_label_str in rctx.attr.cxx_cross_lib.items():
         if target_pair == exec_pair:
             continue
+        target_os = target_pair.split("-")[0]
         cxx_cross_lib_label = Label(cxx_cross_lib_label_str)
         cxx_cross_lib_path = _pkg_path_from_label(cxx_cross_lib_label)
         cxx_cross_lib_include_dirs[target_pair] = [
@@ -165,8 +166,8 @@ def llvm_config_impl(rctx):
         ]
         cxx_cross_lib_link_flags[target_pair] = [
             "-L" + cxx_cross_lib_path + "/lib",
-            "-l:libc++.a",
-            "-l:libc++abi.a",
+            "-l:libc++.a" if target_os != "darwin" else "-lc++",
+            "-l:libc++abi.a" if target_os != "darwin" else "-lc++abi",
         ]
         cxx_cross_lib_cxx_flags[target_pair] = [
             "-nostdinc++",
